### DISCUSSION TITLE
libSkiaSharp.so crash (Sigsegv, Segv_Maperr) on Android only.

### DISCUSSION THREAD

**Discussion by jimiatnymbl** [mono/SkiaSharp#1695] [Q&A]
Background:
Xamarin.Forms app.  Released to both Google Play and Apple.  According to Google Play our crash rate is hovering around 3 - 4% and the majority of the crashes are "within" libSkiaSharp.  Our crash rate for iOS is very, very low.

Of course my first thought was it's our code.  But I find it strange that it only happens on Android (our app is 90% shared code, with very few Custom Renderers etc).

Unfortunately, the only information I have on Google Play is the following signature (across many devices and Android versions):

```
signal 11 (SIGSEGV), code 1 (SEGV_MAPERR)
libSkiaSharp.so (offset 0x1000)

#00  pc 00000000000c626e  /data/app/appid/split_config.armeabi_v7a.apk!lib/armeabi-v7a/libSkiaSharp.so (offset 0x1000)

```

Sometimes it's arm64-v8a, sometimes it's armeabi-v7a, depending on the device of course.  Has anyone else seen this or have any ideas on where to look for more information?

Thanks in advance!

**Comment by glintpursuit** [comment]
@jimiatnymbl 

We are also facing same issue. 

Stack trace is not much helpful 

*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
pid: 0, tid: 0 >>> com.xxxx.xxxxx.app <<<

backtrace:
  #00  pc 00000000000aeea8  /data/app/~~kUvZ0qWNKt5dPFntgMmNQg==/com.xx.xxxxx.app-fOAnxr8cMiayfKKjGIE7Yw==/split_config.arm64_v8a.apk!lib/arm64-v8a/libSkiaSharp.so (offset 0x1000)

content_copy

**Reply by jimiatnymbl** [reply]
As it turns out this was on our end.  We are using SkiaSharp to draw a countdown "circle".  I created a test app that would run this countdown 1000 times and I was able to reproduce the crash locally (turns out it was happening about 1 in 500 iterations).

The issue was a timing issue.  The canvas would be gone but we would have one last OnTimer() event in the queue that was trying to draw the circle.

Since fixing this our Android crash rate is well below 1% now.

**Reply by glintpursuit** [reply]
Ohh Great. 

I am still not sure what is happening with our app, as we don't have stack trace & app center is also not able to catch it.

We are using skia for drawing on image.

**Comment by taublast** [comment]
Have this problem on maui using "SkiaSharp.Skottie" Version="2.88.3" (along with "SkiaSharp.Views.Maui.Controls.Compatibility" Version="2.88.3")
Everything works fine on emulator, on real android device it crashes while invoking Animation.Parse() or TryParse(). Only SkiaSharp.Skottie assembly is affected for me, "standart" skiasharp works like a charm.
It tends to crash 100% of the time if i just run it, and if i debug and step over it's like 75-90% crash probability. What a fun :)

05-04 17:30:57.610 26359 26359 F DEBUG   : ABI: 'arm64'
05-04 17:30:57.610 26359 26359 F DEBUG   : signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0xe20f8970bf
05-04 17:30:57.610 26359 26359 F DEBUG   :     x0  b40000717f7dc290  x1  00000073448787cc  x2  0000000000000000  x3  0000000000000010
05-04 17:30:57.610 26359 26359 F DEBUG   :     x4  0000000000000000  x5  0000000000000000  x6  b40000710f801d28  x7  0000000000000000
05-04 17:30:57.610 26359 26359 F DEBUG   :     x8  00000000000000b4  x9  0000007009d1e4c0  x10 000000004d85fad4  x11 000000007f7dc490
05-04 17:30:57.610 26359 26359 F DEBUG   :     x12 0000ffff00000eff  x13 00000000dfc6add2  x14 0000007009d1e260  x15 0000000000000004
05-04 17:30:57.610 26359 26359 F DEBUG   :     x16 00000070007a7060  x17 00000073447d9d5c  x18 0000007008de8000  x19 b40000717f7dc290
05-04 17:30:57.610 26359 26359 F DEBUG   :     x20 b40000717f7dc490  x21 0000000000000000  x22 b40000e20f8970c0  x23 b40000720f895830
05-04 17:30:57.610 26359 26359 F DEBUG   :     x24 0000000000000000  x25 b40000717f7dc290  x26 0000007009d1e4b8  x27 0000007009d1e658
05-04 17:30:57.610 26359 26359 F DEBUG   :     x28 00000070006e4c73  x29 0000007009d1e570
05-04 17:30:57.610 26359 26359 F DEBUG   :     lr  00000070001c7a40  sp  0000007009d1e490  pc  00000070001c7a8c  pst 0000000080000000
05-04 17:30:57.610 26359 26359 F DEBUG   : backtrace:
05-04 17:30:57.610 26359 26359 F DEBUG   :       #00 pc 000000000018ca8c  /data/app/~~hqorcLirY52kXWB_uPMYmA==/com.xxx.app-6aeAz4C3BoCIWE2lQwqdbA==/lib/arm64/libSkiaSharp.so

**Comment by taublast** [comment]
Just went around the crash: 

this crashes as described above:
`SkottieApi.skottie_animation_make_from_string (json, json.Length) `- where json is a string, my silly guess is maybe on real device the string becomes *sometimes* fragmentated and then direct memory access results in SIGSEGV

this works fine:
`SkottieApi.skottie_animation_make_from_data (ptr, (IntPtr)span.Length)`

so in my specific case the workaround was to drop Animation.TryParse(string..) and start using TryCreate(SKData..)

```
                var bytes = Encoding.UTF8.GetBytes(json);
                var data = SKData.CreateCopy(bytes);
                if (Animation.TryCreate(data, out var animation))
```

Hope this helps.

**Reply by r2d2Proton** [reply]
That worked for me. Need to try other platforms. Thank you @taublast
